harness/determined/exec/launch.py

```diff
@@ -0,0 +1,94 @@
+import copy
+import json
+import logging
+import subprocess
+from typing import Dict
+
+import determined as det
+import determined.common
+from determined.common import constants, storage
+
+
+def mask_config_dict(d: Dict) -> Dict:
+    mask = "********"
+    new_dict = copy.deepcopy(d)
+
+    # checkpoint_storage
+    hidden_checkpoint_storage_keys = ("access_key", "secret_key")
+    try:
+        for key in new_dict["checkpoint_storage"].keys():
+            if key in hidden_checkpoint_storage_keys:
+                new_dict["checkpoint_storage"][key] = mask
+    except KeyError:
+        pass
+
+    return new_dict
+
+
+def launch() -> int:
+    info = det.get_cluster_info()
+    assert info is not None, "must be run on-cluster"
+    assert info.task_type == "TRIAL", f'must be run with task_type="TRIAL", not "{info.task_type}"'
+
+    # Hack: read the full config.  The experiment config is not a stable API!
+    experiment_config = info.trial._config
+
+    # Perform validations
+    debug = experiment_config.get("debug", False)
+    determined.common.set_logger(debug)
+
+    logging.info(
+        f"New trial runner in (container {info.container_id}) on agent {info.agent_id}: "
+        + json.dumps(mask_config_dict(experiment_config))
+    )
+
+    # TODO: this should go in the chief worker, not in the launch layer.  For now, the
+    # DistributedContext is not created soon enough for that to work well.
+    try:
+        storage.validate_config(
+            experiment_config["checkpoint_storage"],
+            container_path=constants.SHARED_FS_CONTAINER_PATH,
+        )
+    except Exception as e:
+        logging.error("Checkpoint storage validation failed: {}".format(e))
+        return 1
+
+    native_enabled = (
+        experiment_config.get("internal", None) is not None
+        and "native" in experiment_config["internal"]
+    )
+
+    # If native is enabled, harness will load from native entrypoint command
+    if native_enabled:
+        from determined.exec import harness
+
+        return harness.main(None)
+
+    entrypoint_cmd = experiment_config.get("entrypoint", None)
+    slots_per_trial = experiment_config.get("resources", {}).get("slots_per_trial", 1)
+
+    if not entrypoint_cmd:
+        logging.error("Entrypoint not specified in experiment config")
+        return 1
+
+    if isinstance(entrypoint_cmd, str):
+        entrypoint_cmd = entrypoint_cmd.split(" ")
+
+    logging.info(f"Launching: {entrypoint_cmd}")
+
+    # Legacy entrypoint (model_def:Trial) detected
+    # Default to horovod if distributed training, else continue in non-distributed training mode
```

Comment:
nit: don't put a comment that says "Legacy entrypoint detected" outside the `if` statement, because at the indentation level you have that statement, you actually haven't detected that yet.  But I like the comment, I think it is a good explanation, just not quite in the right place.